src/url.cpp

```diff
@@ -0,0 +1,137 @@
+#include <algorithm>
+#include <string>
+#include <iterator>
+#include <unordered_map>
+#include <unordered_set>
+#include <iostream>
+
+#include "url.h"
+
+namespace Url
+{
+
+    /* Character classes */
+    const std::string Url::GEN_DELIMS = ":/?#[]@";
+    const std::string Url::SUB_DELIMS = "!$&'()*+,;=";
+    const std::string Url::ALPHA = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+    const std::string Url::DIGIT = "0123456789";
+    const std::string Url::UNRESERVED = Url::ALPHA + Url::DIGIT + "-._~";
+    const std::string Url::RESERVED = Url::GEN_DELIMS + Url::SUB_DELIMS;
+    const std::string Url::PCHAR = Url::UNRESERVED + Url::SUB_DELIMS + ":@";
+    const std::string Url::PATH = Url::PCHAR + "/";
+    const std::string Url::QUERY = Url::PCHAR + "/?";
+    const std::string Url::FRAGMENT = Url::PCHAR + "/?";
+    const std::string Url::USERINFO = Url::UNRESERVED + Url::SUB_DELIMS + ":";
+    const std::string Url::SCHEME =
+        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+-.";
+
+    Url::Url(const std::string& url): port_(0)
+    {
+        size_t position = 0;
+        size_t index = url.find(':');
+        if (index != std::string::npos)
+        {
+            // All the characters in our would-be scheme must be in SCHEME
+            if (url.substr(0, index).find_first_not_of(SCHEME) == std::string::npos)
```

Comment:
That's a good idea about using a `unordered_set` or an array or the like for checking inclusion in a set of characters. We can certainly revisit that.

I did worry about the use of `substr`, but `find_first_not_of` doesn't allow one to provide an upper limit. So the question then becomes whether it's better to potentially search past `index` or to make a copy to limit the extent of the search.

Of course, if we end up making our own `find_first_not_of` using the method you suggest, we could support that mode and do away with the `substr` here, too.